### PR TITLE
Adaptive Load FakeStepController

### PR DESCRIPTION
**Pull request by eric846**
FakeStepController activated with a full TypedExtensionConfig to test plugin registration and config validation.
  - Regardless of state, returns a constant RPS (configurable via the config proto).
  - Reports doom if it has just received a benchmark result with an error from Nighthawk Service.
  - Reports converged if it has just received a benchmark result that received score >0 (manipulate the score using some ScoringFunction and some MetricsPlugin); if the latest result was <=0, reports non-converged. Initially reports non-converged.
  - Config proto supports setting an artificial config validation error.

Partially addresses issue #448.

### CODE REVIEW

**Comment by dubious90** [comment]
@wjuan-AFK please review and assign back to me once done.

File: include/nighthawk/adaptive_load/step_controller.h
Comment (htuch) [inline include/nighthawk/adaptive_load/step_controller.h:69]: Can we avoid multiple implementation inheritance here with some refactoring or changes to is-a vs. has-a relationships? Google/Envoy C++ style discourages multiple implementation inheritance.

Comment (eric846) [reply include/nighthawk/adaptive_load/step_controller.h:69]: Yes, I am removing the shared concrete NullConfigValidator that we were going to inherit from. Instead a plugin factory that doesn't need to validate anything can just explicitly override ValidateConfig() with a body that immediately returns absl::OkStatus(). No other ConfigValidator implementations will need to be shared because they will all be plugin-specific.

Comment (eric846) [reply include/nighthawk/adaptive_load/step_controller.h]: For now, I'm eliminating the multiple inheritance but not shaking up plugin_loader or all the tests written according to these interfaces. Filed #465 to consider whether a refactor will clean it up further, after all these PRs land.

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (wjuan-AFK) [inline test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:41]: What does this have to do with the metrics plugin?

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (wjuan-AFK) [inline test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:143]: Why do we expect these to be equal if it does not write doomed reason after successful benchmark result? Shouldn't it be empty or something?

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc]: Good catch, this was a typo.

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:143]: The local variable doomed_reason starts out with the value "untouched" (line 141) before we call step_controller.IsDoomed(doomed_reason). So the value "untouched" should still be intact instead of getting overwritten somewhere inside IsDoomed.

**Comment by wjuan-AFK** [comment]
LGTM

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (oschaaf) [inline test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:27]: Should we double down with the Message until based comparison to avoid missing edge cases?

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (oschaaf) [inline test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:100]: Maybe create a shared constant for the magic number?

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (oschaaf) [inline test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:133]: Reading this, Ibwonder if a single call to obtain status (converged) or failure (doomed) would simplify things here and for the consing code. Just a thought after recently learning about that construct, please feel free to ignore

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (oschaaf) [inline test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:176]: Make 5 a named/shared constant?

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (oschaaf) [inline test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:164]: "error from nighthawk" might be reduplicated by using a constant?

**Review by oschaaf — CHANGES_REQUESTED** [review]
Looks great, here is my feedback

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (wjuan-AFK) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:133]: I agree with that. And even if you want to keep the IsDoomed call, you can still just have that refer to the status. Especially since Doomed and Converged are mutually exclusive states.

Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:133]: Opened #466

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:164]: Done

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:100]: Done -- keeping these constants local since they have no shared meaning across tests

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:27]: Done

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc]: Done (made another 5 zero)

**Review by oschaaf — APPROVED** [review]
Looks good to me

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (dubious90) [inline test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:149]: I got confused by the same thing as William.  I think this would be clearer, and more accurate to the API of the call if we set this to "", and check that it's still empty after IsDoomed is called.

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (dubious90) [inline test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:172]: Why ASSERT_TRUE rather than EXPECT_TRUE here?  Won't the code after still work?

Also, I honestly think that this and the previous test are essentially testing two very similar parts of the same thing, and we could combine them.  But that's an opinion, so I'll leave it up to you.

Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc:172]: Changed to EXPECT_TRUE.

Deleted the previous test that was a subset of this one.

File: test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc
Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc]: One of the potential bugs this test is guarding against is IsDoomed accidentally writing "" to the variable instead of leaving it untouched, so we would want to fail if it started to do that.

What do you think of this clarified variable name and value?

Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc]: Hmm actually I see what you mean: the caller isn't supposed to have anything important in the variable ahead of time, since it might get overwritten. 

The behavior of not writing the variable when IsDoomed returns false is really not very valuable at all. Maybe the behavior when returning false should be undefined, or IsDoomed should write the variable unconditionally, writing "" if it's returning false.

Comment (eric846) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc]: Opened #482 to rewrite IsDoomed to return absl::Status. The only reason for this weird contract is we had not brought in absl::Status at the time.

Meanwhile the test as it currently stands should explicitly test the ill-conceived spec of this function.

What do you think?

Comment (dubious90) [reply test/adaptive_load/fake_plugins/fake_step_controller/fake_step_controller_test.cc]: Sounds good.  Thanks for taking a look at what could be done.  I like the Status solution.

**Pull request merged**
